### PR TITLE
Check if loop body occured before loopTop and if so unmark alignment

### LINKED ISSUE #91838: JIT: "Mismatch in align instruction" in superpmi-replay
superpmi-replay seems to be failing with "Mismatch in align instruction" when run with JitStressRegs=3:

```scala
[00:26:08] Mismatch in align instruction.

[00:26:08] Containing IG: IG117

[00:26:08] loopHeadPredIG: IG117

[00:26:08] loopHeadIG: IG118

[00:26:08] igInLoop: IG121

[00:26:08] igInLoop->backEdge: IG119

[00:26:08] igInLoop has align instruction for : IG122

[00:26:08] Loop:

[00:26:08] 	IG118

[00:26:08] 	IG119

[00:26:08] 	IG120

[00:26:08] 	IG121

[00:26:08] 	IG122

[00:26:08] 	IG123

[00:26:08] 	IG124

[00:26:08] 	IG125

[00:26:08] 	IG126

[00:26:08] 	IG127

[00:26:08] 	IG128

[00:26:08] 	IG129

[00:26:08] 	IG130

[00:26:08] 	IG131

[00:26:08] 	IG132

[00:26:08] 	IG133

[00:26:08] 	IG134

[00:26:08] 	IG135

[00:26:08] 	IG136

[00:26:08] 	IG137

[00:26:08] 	IG138

[00:26:08] 	IG139

[00:26:08] 	IG140

[00:26:08] 	IG141

[00:26:08] 	IG142

[00:26:08] 	IG143

[00:26:08] 	IG144

[00:26:08] 	IG145

[00:26:08] 	IG146

[00:26:08] 	IG147

[00:26:08] 	IG148

[00:26:08] 	IG149

[00:26:08] 	IG150

[00:26:08] 	IG151

[00:26:08] 	IG152

[00:26:08] 	IG153

[00:26:08] 	IG154

[00:26:08] 	IG155

[00:26:08] 	IG156

[00:26:08] 	IG157

[00:26:08] 	IG158

[00:26:08] 	IG159

[00:26:08] 	IG160

[00:26:08] 	IG161

[00:26:08] 	IG162

[00:26:08] 	IG163

[00:26:08] 	IG164

[00:26:08] 	IG165

[00:26:08] 	IG166

[00:26:08] 	IG167

[00:26:08] 	IG168

[00:26:08] 	IG169

[00:26:08] 	IG170

[00:26:08] 	IG171

[00:26:08] 	IG172

[00:26:08] 	IG173

[00:26:08] 	IG174

[00:26:08] 	IG175

[00:26:08] 	IG176

[00:26:08] 	IG177

[00:26:08] 	IG178

[00:26:08] 	IG179

[00:26:08] 	IG180

[00:26:08] 	IG181

[00:26:08] 	IG182

[00:26:08] 	IG183

[00:26:08] 	IG184

[00:26:08] 	IG185

[00:26:08] 	IG186

[00:26:08] 	IG187

[00:26:08] 	IG188

[00:26:08] 	IG189

[00:26:08] 	IG190

[00:26:08] 	IG191

[00:26:08] 	IG192

[00:26:08] 	IG193

[00:26:08] 	IG194

[00:26:08] 	IG195

[00:26:08] 	IG196

[00:26:08] 	IG197

[00:26:08] 	IG198

[00:26:08] 	IG199

[00:26:08] 	IG200

[00:26:08] 	IG201

[00:26:08] 	IG202

[00:26:08] 	IG203

[00:26:08] 	IG204

[00:26:08] 	IG205

[00:26:08] 	IG206

[00:26:08] 	IG207

[00:26:08] 	IG208

[00:26:08] 	IG209

[00:26:08] 	IG210

[00:26:08] 	IG211

[00:26:08] 	IG212

[00:26:08] 	IG213

[00:26:08] 	IG214

[00:26:08] 	IG215

[00:26:08] 	IG216

[00:26:08] ISSUE: <ASSERT> #530064 D:\a\_work\1\s\src\coreclr\jit\emit.cpp (5883) - Assertion failed 'false' in 'System.Numerics.BigIntegerCalculator:PowCore(System.Span`1[uint],int,System.ReadOnlySpan`1[uint],System.ReadOnlySpan`1[uint],System.Span`1[uint],System.Span`1[uint])' during 'Generate code' (IL size 511; hash 0xd0657e4f; Tier1)

[00:26:08] 
```

Console log: https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-91827-merge-6e8d1e4418334afa89/windows-x64-2/1/console.c2bbff16.log?helixlogtype=result

cc @kunalspathak

### PR DESCRIPTION
**Pull request by kunalspathak**
With optimizations conducted in various phases, we might rearrange the blocks such that loop body appears lexically before the loopTop. As such, we should unmark the alignment of such loops, because when we try to find the size of the loop, we don't find the block that targets back to the loop top. We were previously unmarking such loops, however, it was only if the loop body appears just before the loop top. In the failure case, the loop top is several blocks below the loop body. Updated the code in `placeLoopAlignInstructions` to track if we have already seen a particular `bbNatLoopNum`. If yes, and a block is marked as aligned, that means we have already encountered the loop body block that had the same `bbNatLoopNum`. Unmark such loops. 

Fixes: https://github.com/dotnet/runtime/issues/91838

### CODE REVIEW

**Comment by kunalspathak** [comment]
@BruceForstall

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib @AndyAyersMS

**Review by AndyAyersMS — APPROVED** [review]

File: src/coreclr/jit/compiler.cpp [hunk 1]
```diff
@@ -5276,6 +5276,13 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
     weight_t               minBlockSoFar         = BB_MAX_WEIGHT;
     BasicBlock*            bbHavingAlign         = nullptr;
     BasicBlock::loopNumber currentAlignedLoopNum = BasicBlock::NOT_IN_LOOP;
+    bool                   visitedLoopNum[BasicBlock::MAX_LOOP_NUM];
+    memset(visitedLoopNum, false, sizeof(visitedLoopNum));
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5280]: It's weird to use `memset` to zero a `bool` array, but ok.

File: src/coreclr/jit/compiler.cpp [hunk 2]
```diff
@@ -5276,6 +5276,13 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
...
+
+#ifdef DEBUG
+    int visitedBlockForLoopNum[BasicBlock::MAX_LOOP_NUM];
+    memset(visitedBlockForLoopNum, false, sizeof(visitedBlockForLoopNum));
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5284]: It's weird to use `false` as a value to zero an `int` array.

File: src/coreclr/jit/compiler.cpp [hunk 3]
```diff
@@ -5276,6 +5276,13 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
     weight_t               minBlockSoFar         = BB_MAX_WEIGHT;
     BasicBlock*            bbHavingAlign         = nullptr;
     BasicBlock::loopNumber currentAlignedLoopNum = BasicBlock::NOT_IN_LOOP;
+    bool                   visitedLoopNum[BasicBlock::MAX_LOOP_NUM];
+    memset(visitedLoopNum, false, sizeof(visitedLoopNum));
+
+#ifdef DEBUG
+    int visitedBlockForLoopNum[BasicBlock::MAX_LOOP_NUM];
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5283]: block nums can't be negative; use `unsigned`?

File: src/coreclr/jit/compiler.cpp [hunk 4]
```diff
@@ -5352,12 +5359,19 @@ PhaseStatus Compiler::placeLoopAlignInstructions()
                 madeChanges       = true;
                 unmarkedLoopAlign = true;
             }
-            else if ((block->bbNatLoopNum != BasicBlock::NOT_IN_LOOP) && (block->bbNatLoopNum == loopTop->bbNatLoopNum))
+            else if (visitedLoopNum[loopTop->bbNatLoopNum])
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5362]: Don't you need to check that `block->bbNatLoopNum != BasicBlock::NOT_IN_LOOP`?

Comment (kunalspathak) [reply src/coreclr/jit/compiler.cpp:5362]: Not really. `block` is the previous block before `loopTop` and we want to check here if we have seen a loop body related to `loopTop->bbNatLoopNum`.

Comment (BruceForstall) [reply src/coreclr/jit/compiler.cpp:5362]: Oh, sorry, I meant, don't you need to check `loopTop->bbNatLoopNum != BasicBlock::NOT_IN_LOOP`?

It appears that sometimes it does equal NOT_IN_LOOP, leading to reading beyond the array bounds.

**Pull request merged**